pkg/api/v1beta1/types.go

```diff
@@ -307,6 +307,19 @@ type TypeMeta struct {
 	APIVersion        string    `json:"apiVersion,omitempty" description:"version of the schema the object should have"`
 	Namespace         string    `json:"namespace,omitempty" description:"namespace to which the object belongs; must be a DNS_SUBDOMAIN; 'default' by default"`
 
+	// GenerateName indicates that the name should be made unique by the server prior to persisting
+	// it. A non-empty value for the field indicates the name will be made unique (and the name
+	// returned to the client will be different than the name passed). The value of this field will
+	// be combined with a unique suffix on the server, and takes precedence over the Name field.
+	// The provided value must be valid within the rules for Name, and may be truncated by the length
+	// of the suffix required to make the value unique on the server.
+	//
+	// If this field is specified, the server will NOT return a 409 if Name exists - instead, it will
+	// either return 201 Created or 500 with Reason TryAgainLater indicating a unique name could not
+	// be found in the time allotted, and the client should retry (optionally after the time indicated
+	// in the Retry-After header).
+	GenerateName string `json:"generateName,omitempty" description:"an optional prefix to use to generate a unique name; has the same validation rules as name; optional and overrides name if specified"`
```

Comment:
Continuing discussion from #3908:

I would like to make a hard rule that we don't mutate writable fields the client sends. OTOH, we can default unspecified fields all we want, and should make defaults fully explicit in the metadata and spec.

GET -> patch -> PUT (narrow update) needs to work in a straightforward manner.

Submit/apply/ensure (declarative update), extract (create clean config), and clone need to be based on user intent. The plan of record is to use annotations to record user intent rather than an official second layer of desired state.

Cloning groups of objects would indeed be even harder. However, I don't see that as being a likely scenario for objects with independently uniquified names.